### PR TITLE
[#9494] Consolidate Instructor Privilege RESTFul API

### PR DESCRIPTION
**Pull request by junming403**
Part of #9494 

**Outline of Solution**
Redesigned response DTO for instructor privilege. According to request parameters, selected privileges will be returned. 

There are in total 3 cases to cover all the related old endpoints.
(1). Course id and `INSTRUCTOR_IS_PRIVILEGE_MAP_NEEDED` set to `true`, a map containing privileges for different instructor roles will be returned. related old endpoint: `GetCourseEditDetailsAction`

(2). Course id and `INSTRUCTOR_PRIVILEGES_IS_ALL_NEEDED` has been set to `true`. In this case complete privileges of the instructor will be returned. related old endpoint: `GetInstructorCoursesAction`

(3). Course id (feedback session name is optional here). In this case selected privileges will be returned. related old endpoints: `GetInstructorPrivilegeAction`.

(4). Course id , section name. In this case two special privileges will be returned. related old endpoints: `GetInstructorCourseDetailsAction`

This solution's new response DTO is more general to fit into different requirements. Some front end changes will be required.

### CODE REVIEW

File: src/main/java/teammates/ui/webapi/output/InstructorPrivilegeData.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+import java.util.Map;
+
+import teammates.common.datatransfer.InstructorPrivileges;
+
+/**
+ * The output format of instructor privileges.
+ */
+public class InstructorPrivilegeData extends ApiOutput {
+    private InstructorPrivileges privileges;
```
Comment (xpdavid) [inline src/main/java/teammates/ui/webapi/output/InstructorPrivilegeData.java]: I don't think you should expose `InstructorPrivileges`. It is kind of internal data structure to store privilege of an instructor (i.e. map).

image

This is what you should put in the response DTO. The structure of the DTO will the same while it changed for different granularity (course, section, feedbackSession).

Comment (junming403) [reply src/main/java/teammates/ui/webapi/output/InstructorPrivilegeData.java]: Sorry, you are saying we should wrap the `InstructorPrivileges` to make the response DTO(still a map)? 

I just used `InstructorPrivileges` as  a container, which contains 3 maps(courseLevel, sectionLevel, sessionLevel), and put privileges as required into the corresponding map. 
From my understanding, those privileges are quite self-explanatory from what I saw from the old endpoints. For example,
```
        response.setCanModifyCourse(
                instructor.isAllowedForPrivilege(Const.ParamsNames.INSTRUCTOR_PERMISSION_MODIFY_COURSE));
```
I did not say any translation needed here.

So is it okay to just create my own version of container of 3 maps and use it to substitute `InstructorPrivileges`?

Comment (xpdavid) [reply src/main/java/teammates/ui/webapi/output/InstructorPrivilegeData.java]: Yup. A map is fine but don't use `InstructorPrivileges` for convenience.

**Comment by junming403** [comment]
Updated, replace `InstructorPrivileges` -> `InstructorPrivilegesBundle`. Any suggestions? 
Since `map` is used on the response DTO, we need to agree on the key values of the map. Any preferred way to do that? Or I just create a const file in front end?

**Pull request closed without merging**